Dockerfile

```diff
@@ -1,24 +1,56 @@
-FROM node:latest
+ARG ARCH=amd64
+
+FROM alpine as qemu
+
+RUN apk add --no-cache curl
+
+RUN curl -fsSL https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/qemu-arm-static -O \
+	&& chmod +x qemu-arm-static
+
+RUN curl -fsSL https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/qemu-aarch64-static -O \
+	&& chmod +x qemu-aarch64-static
+
+# ----------------------------------------------------------------------------
+
+FROM ${ARCH}/node:11
+
+COPY --from=qemu qemu-arm-static qemu-aarch64-static /usr/bin/
 
 ENV DEBIAN_FRONTEND noninteractive
-ENV INITRD No
-ENV NODE_ENV production
 
-RUN apt-get update
+RUN apt-get update && apt-get install -yq --no-install-recommends \
+	git \
+	curl \
+	libsystemd-dev \
+	build-essential \
+	libssl-dev \
+	net-tools \
+	ldnsutils \
+	&& apt-get clean \
+	&& rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+WORKDIR /app
+
+COPY package.json ./
+COPY source/ ./source
 
-RUN apt-get install \
-  git \
-  curl \
-  libsystemd-dev \
-  build-essential \
-  libssl-dev \
-  net-tools
+RUN npm install --production
 
-COPY package.json /root/dohnut/
-COPY source/ /root/dohnut/source/
+RUN ln -s /app/source/bin.js /usr/local/bin/dohnut
 
-RUN cd /root/dohnut && npm install
+ARG BUILD_DATE
+ARG BUILD_VERSION
+ARG VCS_REF
 
-RUN ln -s /root/dohnut/source/bin.js /usr/local/bin/dohnut
+LABEL maintainer="kylemharding@gmail.com"
+LABEL org.label-schema.schema-version="1.0"
+LABEL org.label-schema.name="commonshost/dohnut"
+LABEL org.label-schema.description="Dohnut is a DNS to DNS-over-HTTPS (DoH) proxy server"
+LABEL org.label-schema.url="https://help.commons.host/dohnut/"
+LABEL org.label-schema.vcs-url="https://github.com/commonshost/dohnut"
+LABEL org.label-schema.docker.cmd="docker run -p 53:53/tcp -p 53:53/udp commonshost/dohnut --listen 0.0.0.0:53 --doh commonshost"
+LABEL org.label-schema.build-date="${BUILD_DATE}"
+LABEL org.label-schema.version="${BUILD_VERSION}"
+LABEL org.label-schema.vcs-ref="${VCS_REF}"
 
-ENTRYPOINT ["dohnut"]
+ENTRYPOINT [ "dohnut" ]
```

Comment:
I noticed there is also a `HEALTHCHECK` command in Docker. Do we need that? Pi-hole's template uses it, fwiw.

Reply:
Yeah, I tested for a little bit with healthcheck in place, but then I noticed users could bind to any interface:port combination they wanted at runtime so that's kind of hard to test against reliably.

My script actually had to parse the output of `netstat` with `grep` and `awk` to figure out which port node was binding too then hit it with some dns tests. It worked but it wasn't very elegant.

In my experience, if users want to use a healthcheck they will add it to their docker-compose configuration themselves since they know what they want to test for, and it doesn't need to be in the base image.

Obviously up to you though, and if you want help I can show you what I tried.